### PR TITLE
Implement Domain endpoint for compiling templates to schemas

### LINKED ISSUE #57: [FEATURE] Compiling templates into schemas
**Is your feature request related to a problem?**
Adding query functionality to simple-schema requires a schema

**What solution would you like?**
In order to compile templates to schemas, we need to:
- Define a schema request type at the `/object` endpoint per `docs/endpoint.md` to create schemas
- On schema creation, generate indices
- Once a schema is created, a side-effect workflow should be triggered:
  - Create a stateful object for a domain (for now, in a concurrent hashmap)
  - Dynamically register a handler (via URL parameter) for requests to that schema
    - For now, just allow introspective `GET` requests

**What alternatives have you considered?**
N/A

**Do you have any additional context?**
N/A

### PR DESCRIPTION
**Pull request by Swiddis**
### Description
This PR contains the bulk of the routing work for the Domain endpoint, including basic functionality to combine templates into a schema object.
- See [this comment on issue 16](https://github.com/opensearch-project/simple-schema/issues/16#issuecomment-1324413301) for more context on the endpoint
- See `docs/schema-workflow.md` for usage instructions regarding this implementation.

### Issues Resolved
- Resolves #57 
- Is a prerequisite for work on #58 and #45 

### Check List
- [X] New functionality includes testing.
  - [X] All tests pass, including unit test, integration test and doctest
- [X] New functionality has been documented.
  - [ ] New functionality has javadoc added
  - [X] New functionality has user manual doc added
- [X] Commits are signed per the DCO using --signoff

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

File: service/src/main/kotlin/org/opensearch/simpleschema/action/CreateSimpleSchemaDomainRequest.kt [hunk 1]
```diff
@@ -0,0 +1,98 @@
...
+    constructor(input: StreamInput) : super(input) {
+        objectId = input.readString()
+        entities = input.readStringList()
+    }
+
+    /**
+     * {@inheritDoc}
+     */
+    override fun validate(): ActionRequestValidationException? {
+        return null
```
Comment (YANG-DB) [inline service/src/main/kotlin/org/opensearch/simpleschema/action/CreateSimpleSchemaDomainRequest.kt:81]: why is this null ?

File: service/src/main/kotlin/org/opensearch/simpleschema/action/CreateSimpleSchemaDomainResponse.kt [hunk 2]
```diff
@@ -0,0 +1,57 @@
...
+import org.opensearch.common.xcontent.XContentParser
+import org.opensearch.common.xcontent.XContentParserUtils
+import org.opensearch.commons.utils.logger
+import org.opensearch.simpleschema.domain.DomainRepository
+import org.opensearch.simpleschema.model.BaseResponse
+import org.opensearch.simpleschema.model.RestTag
+import java.io.IOException
+
+internal class CreateSimpleSchemaDomainResponse : BaseResponse {
+    private var objectId: String
```
Comment (YANG-DB) [inline service/src/main/kotlin/org/opensearch/simpleschema/action/CreateSimpleSchemaDomainResponse.kt:17]: replace objectId with name (domain name)

File: service/src/main/kotlin/org/opensearch/simpleschema/action/GetSimpleSchemaDomainRequest.kt [hunk 3]
```diff
@@ -0,0 +1,85 @@
...
+import org.opensearch.common.xcontent.XContentParser
+import org.opensearch.common.xcontent.XContentBuilder
+import org.opensearch.common.xcontent.ToXContent
+import org.opensearch.common.xcontent.XContentParserUtils
+import org.opensearch.commons.utils.logger
+import org.opensearch.simpleschema.model.RestTag
+import java.io.IOException
+
+internal class GetSimpleSchemaDomainRequest : ActionRequest, ToXContentObject {
+    val objectId: String
```
Comment (YANG-DB) [inline service/src/main/kotlin/org/opensearch/simpleschema/action/GetSimpleSchemaDomainRequest.kt:17]: same here - objectId => name

File: service/src/main/kotlin/org/opensearch/simpleschema/action/GetSimpleSchemaDomainResponse.kt [hunk 4]
```diff
@@ -0,0 +1,53 @@
...
+import org.opensearch.common.xcontent.XContentParserUtils
+import org.opensearch.commons.utils.logger
+import org.opensearch.simpleschema.model.BaseResponse
+import org.opensearch.simpleschema.model.RestTag
+import java.io.IOException
+
+internal class GetSimpleSchemaDomainResponse : BaseResponse {
+
+
+    private var objectId: String
```
Comment (YANG-DB) [inline service/src/main/kotlin/org/opensearch/simpleschema/action/GetSimpleSchemaDomainResponse.kt:18]: objectId => name

File: service/src/main/kotlin/org/opensearch/simpleschema/action/SimpleSchemaDomainActions.kt [hunk 5]
```diff
@@ -0,0 +1,89 @@
...
+        )
+        if (SimpleSearchIndex.getSimpleSchemaObject(request.objectId) != null) {
+            log.info("attempted to recreate existing schema: ${request.objectId}")
+            throw OpenSearchStatusException(
+                "Creation failed: Schema ID already exists",
+                RestStatus.BAD_REQUEST
+            )
+        }
+        // TODO unaddressed edge case where compilation is successful but storage fails
+        compile(objectDoc, user)
```
Comment (YANG-DB) [inline service/src/main/kotlin/org/opensearch/simpleschema/action/SimpleSchemaDomainActions.kt:43]: needs a more descriptive name and documentation

File: service/src/main/kotlin/org/opensearch/simpleschema/action/SimpleSchemaDomainActions.kt [hunk 6]
```diff
@@ -0,0 +1,89 @@
...
+        UserAccessManager.validateUser(user)
+        val result = SimpleSearchIndex.getSimpleSchemaObject(request.objectId)
+        result ?: throw OpenSearchStatusException(
+            "Schema not found",
+            RestStatus.NOT_FOUND
+        )
+        return GetSimpleSchemaDomainResponse(result.simpleSchemaObjectDoc.objectId)
+    }
+
+    private fun compile(objectDoc: SimpleSchemaObjectDoc, user: User?) {
```
Comment (YANG-DB) [inline service/src/main/kotlin/org/opensearch/simpleschema/action/SimpleSchemaDomainActions.kt:63]: I think this should be an interface method

File: service/src/main/kotlin/org/opensearch/simpleschema/domain/DomainRepository.kt [hunk 7]
```diff
@@ -0,0 +1,21 @@
...
+    var domains: ConcurrentHashMap<String, DomainResource> = ConcurrentHashMap()
+
+    /**
+     * Adds a [domain] to the repository if the name is not already present,
+     * otherwise return the existing domain.
+     *
+     * @return The domain associated with the same name if already present, otherwise null.
+     */
+    fun createDomain(domain: DomainResource): DomainResource? {
+        return domains.putIfAbsent(domain.name, domain)
```
Comment (YANG-DB) [inline service/src/main/kotlin/org/opensearch/simpleschema/domain/DomainRepository.kt:15]: should return some indication if failed creating new domain due to existing one

File: service/src/main/kotlin/org/opensearch/simpleschema/model/SchemaCompilationType.kt [hunk 8]
```diff
@@ -0,0 +1,123 @@
...
+import org.opensearch.common.xcontent.XContentBuilder
+import org.opensearch.common.xcontent.XContentParserUtils
+import org.opensearch.simpleschema.SimpleSchemaPlugin.Companion.LOG_PREFIX
+import org.opensearch.simpleschema.util.fieldIfNotNull
+import org.opensearch.simpleschema.util.logger
+import org.opensearch.simpleschema.util.stringList
+
+/**
+ * This element represents a schema entity which is stored as a document in the DB
+ * The entity has a type, name, catalog, description and the actual GraphQL SDL as content
```
Comment (YANG-DB) [inline service/src/main/kotlin/org/opensearch/simpleschema/model/SchemaCompilationType.kt:18]: has a group of GraphQL components such as : types, mutations, queries, subscriptions

**Review by YANG-DB — CHANGES_REQUESTED** [review]

File: service/src/main/kotlin/org/opensearch/simpleschema/model/SchemaCompilationType.kt (hunk 8 above)
Comment (Swiddis) [reply service/src/main/kotlin/org/opensearch/simpleschema/model/SchemaCompilationType.kt:18]: Added clarification in doc that IDs referencing these go in the entities list, should they be their own objects?

File: docs/schema-workflow.md [hunk 9]
```diff
@@ -0,0 +1,42 @@
...
+
+Entities can be created, updated, modified, and deleted through the `/_plugins/_simpleschema/object` endpoint.
+Creation is done by `POST`ing to this endpoint with a Json object.
+The object must have a `type` of `schemaEntityType`.
+
+```json
+{
+  "schemaEntityType": {
+    "name": "Author",
+    "objectId": "testAuthorObject",
```
Comment (YANG-DB) [inline docs/schema-workflow.md:18]: change objectId to name

File: docs/schema-workflow.md [hunk 10]
```diff
@@ -0,0 +1,42 @@
...
+
+Compiled GraphQL Schemas are stored as `Domain` resources,
+available at the `/_plugins/_simpleschema/domain` endpoint.
+As `Domain`s must be uniquely named, they are given an `objectId` instead of a `name`.
+
+Provide the entity list as a list of `objectId`s to be used in compilation.
+
+```json
+{
+  "objectId": "testSchema",
```
Comment (YANG-DB) [inline docs/schema-workflow.md:38]: change objectId to name

**Review by YANG-DB — APPROVED** [review]

**Pull request merged**